File: go/LLPalindrome/LLPalindrome.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func isPalindrome(head *ListNode) bool {
	return approachII(head)
}

// Time -> 0(N) Space -> 0(N)
func approachI(head *ListNode) bool {
	array := make([]int, 0)

	for head != nil {
		array = append(array, head.Val)
		head = head.Next
	}

	start, end := 0, len(array)-1

	for start < end {
		if array[start] != array[end] {
			return false
		}
		start++
		end--
	}

	return true
}

// Time -> 0(N) Space -> 0(1)
func approachII(head *ListNode) bool {
	count := 0
	curr := head
	for curr != nil {
		curr = curr.Next
		count++
	}

	mid := count / 2

	reverse := head
	for mid > 0 {
		reverse = reverse.Next
		mid--
	}

	if count%2 == 1 {
		reverse = reverse.Next
	}

	var endLL *ListNode
	for reverse != nil {
		tmp := reverse.Next
		reverse.Next = endLL
		endLL = reverse
		reverse = tmp
	}

	for endLL != nil {
		if endLL.Val != head.Val {
			return false
		}
		endLL = endLL.Next
		head = head.Next
	}

	return true
}

File: python/llPalindrome.py
class ListNode:
    def __init__(self, val=0, next=None) -> None:
        self.val = val
        self.next = next


def isPalindrome(head: ListNode | None) -> bool:
    return approachII(head)


# Time -> 0(N) Space -> 0(N)
def approachI(head: ListNode | None) -> bool:
    array = []

    while head is not None:
        array.append(head.val)
        head = head.next

    start, end = 0, len(array) - 1
    while start < end:
        if array[start] != array[end]:
            return False

        start += 1
        end -= 1

    return True


# Time -> 0(N) Space -> 0(1)
def approachII(head: ListNode | None) -> bool:
    count = 0
    curr = head

    while curr is not None:
        count += 1
        curr = curr.next

    mid = count // 2

    reverse = head
    while mid > 0:
        if reverse:  # This will always be true just type checking
            reverse = reverse.next
        mid -= 1

    if count % 2 == 1 and reverse:
        reverse = reverse.next

    endLL = None
    while reverse is not None:
        tmp = reverse.next
        reverse.next = endLL
        endLL = reverse
        reverse = tmp

    while endLL is not None and head:
        if endLL.val != head.val:
            return False

        head = head.next
        endLL = endLL.next

    return True
